### PR TITLE
Tests

### PR DESCRIPTION
**Pull request by Mayank447**
Wrote basic functions for the Dirac Operator class and test functions in real space, spectral space and transform

### CODE REVIEW

**Comment by chillenzer** [comment]
There is exactly a single test passing and the `pyproject.toml` is incomplete to the point that the project is uninstallable and it's quite hard to actually run the tests. Did you run the test before submission?

File: Eigenmodes.py
Comment (chillenzer) [inline Eigenmodes.py:1]: Remove this file. We've started off fresh.

File: README.md
Comment (chillenzer) [inline README.md:1]: This is totally out of date. Remove or update.

File: documentation.md [hunk 1]
```diff
@@ -0,0 +1,78 @@
+# Tutorial 
+
+## Dirac Operator -
```
Comment (chillenzer) [inline documentation.md:3]:
```suggestion
## Dirac Operator
```

File: documentation.md [hunk 2]
```diff
@@ -0,0 +1,78 @@
+# Tutorial 
+
+## Dirac Operator - 
+This tutorial covers the application of a finite-dimensional linear operator using its spectral representation.
+
+## Usage -
```
Comment (chillenzer) [inline documentation.md:6]:
```suggestion
## Usage
```

File: documentation.md [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+
+The linear operator is defined by its spectral representation which is given as a `Spectrum` object.
+
+### Defining the `Spectrum`
+
+Defining a spectral representation (a class) [input - eigenfunctions, corresponding eigenvalues, space]. User gives the spectrum object to the Dirac Operator while initializing the object. [We will be using a linear function representation in place of the matrix representation]
```
Comment (chillenzer) [inline documentation.md:12]: Make this a well-phrased paragraph.

File: documentation.md [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+        
+        elif input_space == 'spectral' and output_space == 'spectral':
+            return input_vector
+        
+        elif input_space == 'spectral' and output_space == 'real':
+            # Perform the inverse discrete Fast Fourier transform to go from spectral to real space
+            return np.fft.ifft(input_vector) * np.sqrt(self.n_real)
+        
+        else:
+            raise ValueError("Unsupported space transformation.")
```
Comment (chillenzer) [inline documentation.md:39]: The `_eigenvalues()` method is missing.

File: documentation.md [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+        # Perform the discrete Fast Fourier transform to go from real to spectral space
+        elif input_space == 'real' and output_space == 'spectral':
+            return np.fft.fft(input_vector) / np.sqrt(self.n_real)
+        
+        elif input_space == 'spectral' and output_space == 'spectral':
+            return input_vector
+        
+        elif input_space == 'spectral' and output_space == 'real':
+            # Perform the inverse discrete Fast Fourier transform to go from spectral to real space
+            return np.fft.ifft(input_vector) * np.sqrt(self.n_real)
```
Comment (chillenzer) [inline documentation.md:36]: There is likely a way to write this more concise. We should consider finding that.

File: documentation.md [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+
+```
+# Example 1D derivative:
+class Derivative1D:
+    def __init__(self, num_lattice_points):
+        self.num_lattice_point = num_lattice_points
+        self.L = 1
+        self.eigenvalues = self._eigenvalue(..)
+
+    def transform(self, input_vector, input_space, output_space):
```
Comment (chillenzer) [inline documentation.md:23]: We should consider passing `input_space` and `output_space` as [enums](https://docs.python.org/3/library/enum.html). The `Spectrum` itself could then define the enum, so that the user can get a quick overview over what this spectrum is capable of.

File: documentation.md [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+            return np.fft.ifft(input_vector) * np.sqrt(self.n_real)
+        
+        else:
+            raise ValueError("Unsupported space transformation.")
+```
+
+### Using the operator
+
+Using the operator is straightforward afterwards.
+Take your `spectral_representation` from the last section and define your input:
```
Comment (chillenzer) [inline documentation.md:45]: We don't define a `spectral_representation` above. Either add that to the example above or to the example below.

File: documentation.md [hunk 8]
```diff
@@ -0,0 +1,78 @@
...
+```
+# Example: Represented in real space.
+# This happens to be an eigenvector:
+input_vector = [...]
```
Comment (chillenzer) [inline documentation.md:49]: Put values here.

File: documentation.md [hunk 9]
```diff
@@ -0,0 +1,78 @@
...
+```
+Then, you can call
+```
+operator = Operator(spectral_representation)
+result = operator.apply_to(input_vector)
+
+# because input_vector is an eigenvector:
+assert result == eigenvalue * input_vector
```
Comment (chillenzer) [inline documentation.md:57]: Make an example folder that contains exactly this code as a script (including all setup that's needed), so the user has a nice starting point for their own application.

File: documentation.md [hunk 10]
```diff
@@ -0,0 +1,78 @@
...
+```
+to get the result.
+
+Explain what the code does!
+
+# Testing -
+assert expected == actual
```
Comment (chillenzer) [inline documentation.md:64]: Remove.

**Review by chillenzer — CHANGES_REQUESTED** [review]
Good work! Thank you!

I've put in my first charge of comments. Didn't look at the test yet.

File: pseudospectral/spectrum.py [hunk 11]
```diff
@@ -0,0 +1,63 @@
+import numpy as np
+import scipy.fft
```
[the remark this answers is not in the document]
Comment (Mayank447) [reply pseudospectral/spectrum.py:2]: According to numpy documentation (https://numpy.org/devdocs/reference/routines.fft.html). The SciPy module scipy.fft is a more comprehensive superset of numpy.fft, which includes only a basic set of routines

File: README.md
Comment (Mayank447) [reply README.md:1]: I will update

File: pseudospectral/spectrum.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (chillenzer) [reply pseudospectral/spectrum.py:2]: I presume we won't need that generality but okay.

File: test/test_real_space.py [hunk 12]
```diff
@@ -0,0 +1,91 @@
...
+import pytest
+
+def test_application_to_a_single_eigenfunction(L=4, n=4):
+    """
+    Python test function to test the application of the Dirac operator to a single eigenfunction in real space.
+    """
+    arbitrary_index = 1 # Index of the eigenfunction to be tested
+    operator = DiracOperator(Derivative1D(num_lattice_points=n, L=L))
+    
+    eigenfunction = operator.spectrum.eigenfunction(arbitrary_index)(np.arange(L))
```
Comment (chillenzer) [inline test/test_real_space.py:14]:
```suggestion
    eigenfunction = operator.spectrum.eigenfunction(arbitrary_index)(np.arange(n)/n*L)
```
You're missing your length parameter here. This is why I advocated for having it set to one in the first place! Please consider again to remove this feature for now. We can add it back when we need it.

Scrolling through the rest I see that change must, of course, be made everywhere.

File: test/test_real_space.py [hunk 13]
```diff
@@ -0,0 +1,91 @@
...
+from pseudospectral import DiracOperator, Derivative1D
+import numpy as np
+import pytest
+
+def test_application_to_a_single_eigenfunction(L=4, n=4):
+    """
+    Python test function to test the application of the Dirac operator to a single eigenfunction in real space.
+    """
+    arbitrary_index = 1 # Index of the eigenfunction to be tested
+    operator = DiracOperator(Derivative1D(num_lattice_points=n, L=L))
```
Comment (chillenzer) [inline test/test_real_space.py:12]: We should use [fixtures](https://docs.pytest.org/en/latest/explanation/fixtures.html) for providing `arbitrary_index` and `spectrum` (separately). That way we have a more concise, easier-to-read test and a very straightforward way to generalise this test to other spectra/indices. Of course, this applies to every test.

File: test/test_real_space.py [hunk 14]
```diff
@@ -0,0 +1,91 @@
...
+    
+    eigenfunction = operator.spectrum.eigenfunction(arbitrary_index)(np.arange(L))
+    expected = eigenfunction * operator.spectrum.eigenvalues[arbitrary_index]
+    result = operator.apply_to(eigenfunction, input_space="real", output_space="real")
+    
+    # print("Expected: ", expected, '\n')
+    # print("Result: ", result, '\n')
```
Comment (chillenzer) [inline test/test_real_space.py:19]: Remove! These and all others (I won't tag them all).

File: test/test_real_space.py [hunk 15]
```diff
@@ -0,0 +1,91 @@
...
+#
+# Eigenfunctions are productions of plane waves in each direction
+# In a finite box with (anti-periodic boundary conditions), eigenvalues are linearly spaced with an offset depending on the boundary conditions.
+
+# def main():
+#     # test_application_to_a_single_eigenfunction(L=4, n=4)
+#     test_application_to_superposition_of_two_eigenfunctions(L=5, n=5)
+
+# if __name__ == "__main__":
+#     main()
```
Comment (chillenzer) [inline test/test_real_space.py:91]: Remove! You can have those in your local branch as a reminder if you prefer but not in mainline.

File: test/test_spectral_space.py [hunk 16]
```diff
@@ -0,0 +1,76 @@
...
+
+def test_application_to_a_single_eigenfunction(L=4, n=4):
+    """
+    Python test function to test the application of the Dirac operator to a single eigenfunction in real space.
+    """
+    arbitrary_index = 1 # Index of the eigenfunction to be tested
+    operator = DiracOperator(Derivative1D(num_lattice_points=n, L=L))
+    
+    eigenfunction = np.zeros(L)
+    eigenfunction[arbitrary_index] = 1
```
Comment (chillenzer) [inline test/test_spectral_space.py:16]:
```suggestion
    eigenfunction = np.eye(n)[arbitrary_index]
```
There is another mistake concerning `n` vs. `L`. Please remove this feature as it is obviously a source of distraction from our real task.

File: test/test_spectral_space.py [hunk 17]
```diff
@@ -0,0 +1,76 @@
...
+
+def test_application_to_superposition_of_two_eigenfunctions(L=47, n=47):
+    """
+    Python test function to test the application of the Dirac operator to a superposition of two eigenfunctions.
+    """
+    arbitrary_index = [1,2] # Index of the eigenfunction to be tested
+    operator = DiracOperator(Derivative1D(num_lattice_points=n, L=L))
+    
+    eigenfunction_1, eigenfunction_2 = np.zeros(L), np.zeros(L)
+    eigenfunction_1[arbitrary_index[0]], eigenfunction_2[arbitrary_index[1]] = 1, 1
```
Comment (chillenzer) [inline test/test_spectral_space.py:34]:
```suggestion
    eigenfunction_1, eigenfunction_2 = np.eye(n)[arbitrary_index]
```

File: test/test_spectral_space.py [hunk 18]
```diff
@@ -0,0 +1,76 @@
...
+
+
+# you can deduplicate test code via @pytest.parametrize() (or something like that...)
+
+if __name__ == '__main__':
+    # test_application_to_a_single_eigenfunction()
+    test_application_to_superposition_of_two_eigenfunctions()
+    # test_reports_correct_eigenvalue()
+    # test_eigenvalues()
+    # print("All tests passed!")
```
Comment (chillenzer) [inline test/test_spectral_space.py:76]: As above.

File: test/test_transform.py [hunk 19]
```diff
@@ -0,0 +1,114 @@
...
+import pytest
+
+
+def test_transforms_from_real_to_spectral_space(L=4, n=4):
+    """
+    Python test function to test the transformation of an eigenvector from real space to spectral space.
+    """
+    operator = DiracOperator(Derivative1D(L, n))
+    arbitrary_index = 1
+    array = np.arange(L)
```
Comment (chillenzer) [inline test/test_transform.py:12]: As above.

File: test/test_transform.py [hunk 20]
```diff
@@ -0,0 +1,114 @@
...
+    eigenfunction = operator.spectrum.eigenfunction(arbitrary_index)(array)
+    result = operator.spectrum.transform(eigenfunction, input_space="real", output_space="spectral")
+    
+    # Create spectral coefficients with a single component
+    expected = np.zeros(L)
+    expected[arbitrary_index] = 1.0
```
Comment (chillenzer) [inline test/test_transform.py:18]: As above.

File: test/test_transform.py [hunk 21]
```diff
@@ -0,0 +1,114 @@
+from pseudospectral import DiracOperator, Derivative1D
```
Comment (chillenzer) [inline test/test_transform.py:1]: We don't need an operator here. This is a property of the spectrum.
```suggestion
from pseudospectral import Derivative1D
```
and write all the tests using the spectrum itself.

File: test/test_transform.py [hunk 22]
```diff
@@ -0,0 +1,114 @@
...
+    print("Result: ", result, '\n')
+    assert np.isclose(expected, result).all()
+
+
+
+def test_transforms_multiple_components_from_real_to_spectral_space(L=4, n=4):
+    """
+    Python test function to test the transformation of real coefficients with multiple components from real space to spectral space.
+    """
+    operator = DiracOperator(Derivative1D(L, n))
```
Comment (chillenzer) [inline test/test_transform.py:29]: There a duplication. Remove!

File: test/test_transform.py [hunk 23]
```diff
@@ -0,0 +1,114 @@
...
+#     # test_transforms_from_real_to_spectral_space(L=4, n=4)
+#     # test_transforms_multiple_components_from_real_to_spectral_space(L=4, n=4)
+#     # test_transforms_from_spectral_to_real_space(L=4, n=4)
+#     # test_transforms_multiple_components_from_spectral_to_real_space(L=4, n=4)
+#     pass
+
+# if __name__ == "__main__":
+#     main()
+
+
```
Comment (chillenzer) [inline test/test_transform.py:114]: As above.

**Review by chillenzer — CHANGES_REQUESTED** [review]
~~Also: The tests still don't pass.~~

Sorry, might have forgotten to pull the latest version.

File: Eigenmodes.py
Comment (Mayank447) [reply Eigenmodes.py:1]: Done

File: pseudospectral/dirac_operator.py [hunk 24]
```diff
@@ -0,0 +1,42 @@
+#!/usr/bin/env python3
+
+import numpy as np
+from pseudospectral.spectrum import Derivative1D
+
+class DiracOperator:
```
[the remark this answers is not in the document]
Comment (Mayank447) [reply pseudospectral/dirac_operator.py:6]: I am not sure but in future we may use some properties of Dirac Operator in the class. In that case it won't be justified to call this Linear operator since Dirac Operators are special cases of Linear Operators

File: test/.cph/.test_transform.py_d2d6822c808b8165c140db0a7be46aff.prob
[the remark this answers is not in the document]
Comment (Mayank447) [reply test/.cph/.test_transform.py_d2d6822c808b8165c140db0a7be46aff.prob:1]: Done

File: test/test_real_space.py (hunk 12 above)
Comment (Mayank447) [reply test/test_real_space.py:14]: I set n=L as you suggested. So therefore it boils down to my formula

File: pseudospectral/dirac_operator.py (hunk 24 above)
[the remark this answers is not in the document]
Comment (chillenzer) [reply pseudospectral/dirac_operator.py:6]: No, we won't. If we do so, we'll build something more specific on top of this class.

File: test/test_real_space.py (hunk 12 above)
Comment (chillenzer) [reply test/test_real_space.py:14]: Well, the important thing is not so much if it works (right now) or not but if it is semantically correct. And this isn't...

But on a more practical note, I'm aware that this test works for the default parameters. But you've made them function arguments, so someone could come along and pass something else. Then, this would be wrong.

File: test/test_transform.py (hunk 21 above)
Comment (Mayank447) [reply test/test_transform.py]: Done

File: test/test_real_space.py (hunk 13 above)
Comment (Mayank447) [reply test/test_real_space.py]: Implemented

File: pseudospectral/derivative_1D.py
Comment (chillenzer) [inline pseudospectral/derivative_1D.py:1]: Create a folder `spectra` and move both implemented spectra there.

File: pseudospectral/dirac_operator.py (hunk 24 above)
[the remark this answers is not in the document]
Comment (chillenzer) [reply pseudospectral/dirac_operator.py:5]: Offline discussion: Do it later (see #4).

[the remark this answers is not in the document]
Comment (chillenzer) [reply pseudospectral/dirac_operator.py:5]: Offline discussion: Do it later (see https://github.com/Mayank447/Pseudo-Spectral-Discretization/issues/4).

**Review by chillenzer — CHANGES_REQUESTED** [review]
The code is starting to get into a great shape. Thanks for your good work!

Two tiny things in the comments. Other than that we'll split up the PR and then it is ready to merge.
